### ISSUE TITLE
How to custom docker swarm data path port during installation?

### ISSUE THREAD

**Issue by lastingman** [caprover/caprover#743]
Caprover init docker swarm during installation. Is there any way to custom docker swarm data path port?

Because some cloud providers save port 4789/udp for VPC network, it causes docker swarm nodes can't communicate with each other through default port 4789.

Manually, it can be done with `docker swarm init --data-path-port 5789` to custom the port is used by docker swarm.

**Comment by githubsaturn** [comment]
You can manually init Docker Swarm with any customization you want. Then create `captain-captain` service manually.
Just make sure to add:
-  env var `IS_CAPTAIN_INSTANCE=1`
- Node ID (placement) must be the main leader ID (otherwise after a crash your service may move to another node which completely breaks your setup)
- Volume mount: `/var/run/docker.sock:/var/run/docker.sock`
- Volume mount: /captain:/captain
- Port 3000:3000

Something like this should work
```
mkdir -p /captain

docker service create \
--name captain-captain \
--env IS_CAPTAIN_INSTANCE=1 \
--constraint node.id==`docker node ls --quiet` \
-p 3000:3000 \
-v /var/run/docker.sock:/var/run/docker.sock \
-v /captain:/captain \
caprover/caprover
```

**Comment by lastingman** [comment]
After I start the captain service, it is keeping restart because of can't find containerId for captain-certbot.

docker version: Docker version 19.03.12, build 48a66213fe

```
# docker service logs captain-captain --follow
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    |
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | > caprover@0.0.0 start /usr/src/app
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | > node ./built/server.js
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    |
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | Captain Starting ...
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | July 6th 2020, 3:35:05.910 am    Emptying generated and temp folders.
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | July 6th 2020, 3:35:05.917 am    Ensuring directories are available on host. Started.
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | July 6th 2020, 3:35:05.924 am    Ensuring directories are available on host. Finished.
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | July 6th 2020, 3:35:05.930 am    Network captain-overlay-network is already attached to service: captain-captain
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | July 6th 2020, 3:35:05.936 am    captain-captain (service) has already been connected to secret: captain-salt
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | July 6th 2020, 3:35:05.938 am    Migration not needed, skipping.
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | July 6th 2020, 3:35:05.946 am    Copying fake certificates...
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | July 6th 2020, 3:35:05.949 am    Updating Load Balancer - Setting up NGINX conf file...
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | July 6th 2020, 3:35:05.949 am    Locking NGINX configuration reloading...
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | July 6th 2020, 3:35:05.979 am    SUCCESS: UNLocking NGINX configuration reloading...
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | July 6th 2020, 3:35:05.981 am    Captain Nginx is already running..
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | July 6th 2020, 3:35:05.984 am    Updating NGINX service...
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | July 6th 2020, 3:35:05.994 am    Waiting for 5 seconds for nginx reload to take into effect
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | July 6th 2020, 3:35:10.999 am    NGINX is fully set up and working...
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | July 6th 2020, 3:35:11.002 am    Captain Certbot is already running..
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | July 6th 2020, 3:35:11.006 am    Updating Certbot service...
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | July 6th 2020, 3:35:14.022 am    Retrying to get containerId for captain-certbot retry count:0
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | July 6th 2020, 3:35:17.026 am    Retrying to get containerId for captain-certbot retry count:1
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | July 6th 2020, 3:35:20.034 am    Retrying to get containerId for captain-certbot retry count:2
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | July 6th 2020, 3:35:23.038 am    Retrying to get containerId for captain-certbot retry count:3
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | July 6th 2020, 3:35:26.046 am    Retrying to get containerId for captain-certbot retry count:4
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | July 6th 2020, 3:35:29.054 am    Retrying to get containerId for captain-certbot retry count:5
captain-captain.1.t1iwjb6op4oy@iZbp10xg0kwnf4fuwslwpeZ    | July 6th 2020, 3:35:29.056 am    executeCommand Container: captain-certbot certbot certificates --non-interactive
```

**Comment by githubsaturn** [comment]
@lastingman I don't see any "restart" happening here. I can see that it took 5 retries to get the container ID, but it didn't get restarted. Can you include more lines of logs please?

**Comment by 16hands** [comment]
Hi @githubsaturn 

Thank you for a really wonderful system, it's made the world of difference for our client who uses it (who can't afford US enterprise like systems) 

We are a solely AWS provider, its not possible to use DigitalOcean. Almost exactly as your docs surmise, our client outgrew the one big server setup and we setup caprover as a cluster with the leader as the registry (ECR didn't like it).

All went well until we tried to connect to an app on a worker node, it constantly timed out. Even from within the nginx container it could ping the worker, but not get data from the worker.

This issue from @lastingman give me the clue I needed, and using your suggestion above I got it working

I found that I needed to start completely clean, so I removed the /captain folder (there wasn't anything important in there yet)

then setup the swarm 

`docker swarm init --data-path-port 5789` -- I'm guessing the port number isn't important so long as it different to the AWS flow logs port 

then added caprover:
`mkdir /captain`  (required for the bind below)

``docker service create --name captain-captain --env IS_CAPTAIN_INSTANCE=1 --constraint node.id==`docker node ls --quiet` -p 3000:3000 --mount 'type=bind,src=/var/run/docker.sock,dst=/var/run/docker.sock' --mount 'type=bind,src=/captain,dst=/captain' caprover/caprover`` 

note the use of --mount, v 19.03.12 of docker didn't like the volume (-v) args

Then I ran the caprover setup, and added the worker node on the command line with the commands from the cluster help in the caprover frontend 

and then after adding an App too the worker node it all worked (so good!) 

Thank you again for a truly helpful project
